src/main/java/io/hops/hopsworks/expat/migrations/featurestore/statistics/StatisticsMigration.java

```diff
@@ -253,15 +266,34 @@ private void migrateFeatureDescriptiveStatistics(ResultSet fdsResultSet)
             windowEndCommitTime = commitTime; // for non-time-travel-enabled fgs, set end window as committime
           }
           // migrate fg stats
-          migrateFeatureGroupStatistics(statisticsId, filePath, windowStartCommitTime, windowEndCommitTime,
-            insertFdsStmt, insertFgFdsStmt);
-          // set window start commit time if time travel-enabled fg
-          if (updateFeatureGroupStatisticsCommitWindow(updateFgsStmt, statisticsId, windowStartCommitTime)) {
-            updateFgStatistics = true;
+          boolean fdsInserted = migrateFeatureGroupStatistics(statisticsId, filePath, windowStartCommitTime,
+            windowEndCommitTime, insertFdsStmt, insertFgFdsStmt);
+          
+          if (fdsInserted) {
+            // set window start commit time if time travel-enabled fg
+            if (updateFeatureGroupStatisticsCommitWindow(updateFgsStmt, statisticsId, windowStartCommitTime)) {
+              updateFgStatistics = true;
+            }
+          } else {  // if fds not inserted
+            // this fg statistics could not be migrated to DB, so we delete the fg stats row.
+            LOGGER.info(String.format(
+              "[migrateFeatureDescriptiveStatistics] -- marking fg statistics for deletion, with id '%s' and " +
+                "feature group id '%s'", statisticsId, entityId));
+            deleteFeatureGroupStatistics(deleteFgsStmt, statisticsId);
+            deleteFgStatistics = true;
           }
         } else if (entityType.equals(TRAINING_DATASET)) {
-          migrateTrainingDatasetStatistics(statisticsId, filePath, commitTime, insertFdsStmt, insertTrainDatasetFdsStmt,
-            insertTestDatasetFdsStmt, insertValDatasetFdsStmt);
+          boolean fdsInserted = migrateTrainingDatasetStatistics(statisticsId, filePath, commitTime, insertFdsStmt,
+            insertTrainDatasetFdsStmt, insertTestDatasetFdsStmt, insertValDatasetFdsStmt);
+          
+          if (!fdsInserted) {   // if fds not inserted
+            // this td statistics could not be migrated to DB, so we delete the td stats row.
+            LOGGER.info(String.format(
+              "[migrateFeatureDescriptiveStatistics] -- marking td statistics for deletion, with id '%s' and " +
+                "training dataset id '%s'", statisticsId, entityId));
+            deleteTrainingDatasetStatistics(deleteTdsStmt, statisticsId);
```

Comment:
Similar question here. If we failed to insert, what do we delete here?

Reply:
Same answer than before, but for the training dataset statistics table.